web3/providers/websocket/websocket_v2.py

```diff
@@ -160,71 +160,46 @@ async def make_request(self, method: RPCEndpoint, params: Any) -> RPCResponse:
-                        "from the cache."
-                    )
-                    return self._request_processor.pop_raw_response(
-                        cache_key=request_cache_key,
-                    )
-
-                else:
-                    if not self._ws_lock.locked():
-                        async with self._ws_lock:
-                            self.logger.debug(
-                                f"Response for id {request_id} is not cached, calling "
-                                "`recv()` on websocket."
-                            )
-                            try:
-                                # keep timeout low but reasonable to check both the
-                                # cache and the websocket connection for new responses
-                                response = await self._ws_recv(timeout=0.5)
-                            except asyncio.TimeoutError:
-                                # keep the request timeout around the whole of this
-                                # while loop in case the response sneaks into the cache
-                                # from another call.
-                                continue
-
-                        response_id = response.get("id")
-
-                        if response_id == request_id:
-                            self.logger.debug(
-                                f"Received and returning response for id {request_id}."
-                            )
-                            return response
-                        else:
-                            # cache all responses that are not the desired response
-                            self.logger.debug("Undesired response received, caching.")
-                            is_subscription = (
-                                response.get("method") == "eth_subscription"
-                            )
-                            self._request_processor.cache_raw_response(
-                                response, subscription=is_subscription
-                            )
+    async def _get_response_for_request_with_id(self, request_id: RPCId) -> RPCResponse:
+        # check cache early to avoid unnecessary websocket recv()
+        cache_key = generate_cache_key(request_id)
+        cached_response = self._request_processor.pop_raw_response(cache_key)
+        if cached_response is not None:
+            return cached_response
 
+        # if not in cache, kick off two competing tasks to wait for the desired response
         try:
-            # Add the request timeout around the while loop that checks the request
-            # cache and tried to recv(). If the request is neither in the cache, nor
-            # received within the request_timeout, raise ``TimeExhausted``.
-            return await asyncio.wait_for(
-                _match_response_id_to_request_id(), self.request_timeout
+            completed, pending = await asyncio.wait_for(
+                asyncio.wait(
```

Comment:
Any chance this could be extracted to a utility for DRYness? Similar `asyncio.wait` happening in [web3/manager.py](https://github.com/ethereum/web3.py/pull/3176/files#diff-c902d776bb4f676fa5cca4ee81273cd66d1113db8109d4f607167681b0b8bdeaR376).
Maybe could pass an argument for the `request_id` vs `subscription`?

Reply:
Good point. Not sure where that could reside 🤔. Any ideas?

Reply:
It may make sense in a `web3/providers/websocket/utils.py` or something similar. I'll toy with that idea tomorrow.

Reply:
I noticed references to `web3._utils` as well if we want to consider creating `websockets.py` within that. Can't tell if using `web3/providers/websocket/utils.py` would be a problem if using it within other modules. Keeping it private is probably better though.